src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/LinkedCasesMapping.java

```diff
@@ -20,17 +23,66 @@ private LinkedCasesMapping() {
-                .map(CaseLinkDetails::getCaseReference)
-                .filter(StringUtils::isNotBlank)
-                .map(caseReference -> ServiceLinkedCases.builder()
-                    .caseReference(caseReference)
-                    .caseName(HearingsCaseMapping.getPublicCaseName(caseData))
-                    .reasonsForLink(HearingsCaseMapping.getReasonsForLink(caseData))
-                    .build())
-                .collect(Collectors.toList());
+            .orElseGet(Collections::emptyList).stream()
+            .filter(Objects::nonNull)
+            .map(CaseLink::getValue)
+            .filter(Objects::nonNull)
+            .map(CaseLinkDetails::getCaseReference)
+            .filter(StringUtils::isNotBlank)
+            .map(caseReference -> ServiceLinkedCases.builder()
+                .caseReference(caseReference)
+                .caseName(HearingsCaseMapping.getPublicCaseName(caseData))
+                .reasonsForLink(HearingsCaseMapping.getReasonsForLink(caseData))
+                .build())
+            .collect(Collectors.toList());
+    }
+    
+    @SuppressWarnings("PMD")
+    public static List<ServiceLinkedCases> getLinkedCasesWithNameAndReasons(SscsCaseData caseData, CcdCaseService ccdCaseService) throws GetCaseException {
+        List<String> linkedReferences = Optional.ofNullable(caseData.getLinkedCase())
+            .orElseGet(Collections::emptyList).stream()
+            .filter(Objects::nonNull)
+            .map(CaseLink::getValue)
+            .filter(Objects::nonNull)
+            .map(CaseLinkDetails::getCaseReference)
+            .collect(Collectors.toList());
+
+        if (linkedReferences.isEmpty()) {
+            return Collections.emptyList();
+        }
+
+        List<ServiceLinkedCases> serviceLinkedCases = new ArrayList<>();
+        ServiceLinkedCases linkedCase;
+        for (String linkRef : linkedReferences) {
+            if (linkRef != null) {
+                //suppressed warning
+                linkedCase = new ServiceLinkedCases();
+                linkedCase.setCaseReference(linkRef);
+                linkedCase.setCaseName(HearingsCaseMapping.getPublicCaseName(getLinkedCaseData(
+                    caseData,
+                    ccdCaseService,
+                    linkRef
+                )));
+                linkedCase.setReasonsForLink(HearingsCaseMapping.getReasonsForLink(caseData));
+                serviceLinkedCases.add(linkedCase);
+            }
+        }
+        return serviceLinkedCases;
+    }
+
+    private static SscsCaseData getLinkedCaseData(SscsCaseData caseData, CcdCaseService ccdCaseService, String caseReference) throws GetCaseException {
+        List<CaseLink> caseLink = caseData.getLinkedCase();
+
+        String linkedReference = null;
+
+        for (CaseLink link : caseLink) {
+            if (link != null && link.getValue() != null && link.getValue().getCaseReference() != null && link.getValue().getCaseReference().equals(caseReference)) {
```

Comment:
I'd hope none of these could be null but it would be nicer to use an Optional.ofNullable here too